### PR TITLE
Feat/number input

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
New `NumberInput` component.
`locale` works on Android but not on iOS, our theory is that it's related to RN-intl issues with hermes and it should be solved in RN70. For now we are not supporting it (no API, commented out in types).
WOAUILIB-2640

## Changelog
New `NumberInput` component

### CODE REVIEW

**Review by lidord-wix — COMMENTED** [review]
Before reviewing the code, I found some weird issues, not sure that all of them are blockers but maybe it's not a big deal to fix them: 
- once you get to the infinity sign (did you know it's possible? just enter many `1` until you get it), you can't remove the infinity sign with a backspace.
- I can't paste a number into the field.
- when I mark all the input to delete it, it only deletes the last char.
- once you add any digit to `999,999,999,999,999` it becomes `10,000,000,000,000,000.` can we even control it or it's the format?
- when entering 855...555 many times, it adds digits that are not 5, and when pressing the backspace couple of times it changes the value to `855.56.` for some reason.
- It lets you add a `.` after the number - try to write `1.9.`
- Android - the input isn't vertically aligned with the prefix.

Let me know if something is not clear

**Comment by M-i-k-e-l** [comment]
> * once you get to the infinity sign (did you know it's possible? just enter many `1` until you get it), you can't remove the infinity sign with a backspace.

I don't think this is an interesting case (extreme edge case), but since it is blocking the UI I'll try to fix it.

> * I can't paste a number into the field.
> * when I mark all the input to delete it, it only deletes the last char.

I'll add `contextMenuHidden` - you cannot paste or mark anymore.

> * once you add any digit to `999,999,999,999,999` it becomes `10,000,000,000,000,000.` can we even control it or it's the format?
> * when entering 855...555 many times, it adds digits that are not 5, and when pressing the backspace couple of times it changes the value to `855.56.` for some reason.

I don't think these are interesting cases (extreme edge case), I believe this comes from the way numbers are represented in the computer which has limited accuracy. It might be possible to improve it using some libraries, but I don't think it's useful for our apps (it's more important in scientific and (perhaps even more) in financial programs).

> * It lets you add a `.` after the number - try to write `1.9.`

Weird, I thought I handled it, I'll take a look

> * Android - the input isn't vertically aligned with the prefix.

I think this is in our `TextField`, but I'll take a look

**Comment by M-i-k-e-l** [comment]
@lidord-wix 
I think there's a problem with the style (the Android issue you've found), I might need to look at this with a UX.

**Comment by M-i-k-e-l** [comment]
> @lidord-wix I think there's a problem with the style (the Android issue you've found), I might need to look at this with a UX.

@lidord-wix I've adjusted the code, it'll allow for more flexibility and fixes the screen
Please ignore the failing build, we have some issues with builds ATM.

**Review by lidord-wix — COMMENTED** [review]
I finally had the chance to review it - sorry for the delay 🙏🏽 
I left some comments :)

File: src/components/numberInput/Presenter.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import {Constants} from '../../commons/new';
+
+export interface Data {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/Presenter.ts:3]: This is an internal interface.

File: src/components/numberInput/Presenter.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  value: number;
+  endsWithDecimalSeparator: boolean;
+  formattedNumber: string;
+  maxLength: number;
+}
+
+/**
+ * `undefined` is a legitimate result (replace current data with it), while `null`
+ * means data should not be set (error state or previous data is already `undefined`)
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/Presenter.ts:12]: The input for the function, it's called `currentData?: Data` while we return `newData` which can also be `undefined` or `null`

File: src/components/numberInput/Presenter.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
+import {Constants} from '../../commons/new';
+
+export interface Data {
+  value: number;
+  endsWithDecimalSeparator: boolean;
+  formattedNumber: string;
+  maxLength: number;
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/Presenter.ts:7]: No, that's the `fractionDigits`, this is the total allowed string length.
This is how we avoid writing the text before we validate it.

File: src/components/numberInput/Presenter.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    newNumber = Number(removeLastChar(newInput));
+  } else {
+    newNumber = Number(newInput);
+  }
+  if (isNaN(newNumber)) {
+    return null;
+  }
+
+  const endsWithDecimalSeparator =
+    newNumber.toString().length !== newInput.length || (newInput.length === 1 && _isLastCharDecimalSeparator);
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/Presenter.ts:84]: Yes...
This solves an edge case with a different locale (`de`, you can change the code and run the tests).
Do you think it's better with `let endsWithDecimalSeparator` instead?
